x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/math/BinaryOptionalNumericFunction.java

```diff
@@ -0,0 +1,127 @@
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.sql.expression.function.scalar.math;
+
+import org.elasticsearch.xpack.sql.expression.Expression;
+import org.elasticsearch.xpack.sql.expression.Expressions;
+import org.elasticsearch.xpack.sql.expression.Expressions.ParamOrdinal;
+import org.elasticsearch.xpack.sql.expression.function.scalar.ScalarFunction;
+import org.elasticsearch.xpack.sql.expression.function.scalar.math.BinaryOptionalMathProcessor.BinaryOptionalMathOperation;
+import org.elasticsearch.xpack.sql.expression.gen.pipeline.Pipe;
+import org.elasticsearch.xpack.sql.expression.gen.script.ScriptTemplate;
+import org.elasticsearch.xpack.sql.tree.Source;
+import org.elasticsearch.xpack.sql.type.DataType;
+
+import java.util.Arrays;
+import java.util.Locale;
+import java.util.Objects;
+
+import static java.lang.String.format;
+import static org.elasticsearch.xpack.sql.expression.TypeResolutions.isNumeric;
+import static org.elasticsearch.xpack.sql.expression.gen.script.ParamsBuilder.paramsBuilder;
+
+public abstract class BinaryOptionalNumericFunction extends ScalarFunction {
+
+    private final Expression left, right;
+    
+    public BinaryOptionalNumericFunction(Source source, Expression left, Expression right) {
+        super(source, right != null ? Arrays.asList(left, right) : Arrays.asList(left));
+        this.left = left;
+        this.right = right;
+    }
+    
+    @Override
+    protected TypeResolution resolveType() {
+        if (!childrenResolved()) {
+            return new TypeResolution("Unresolved children");
+        }
+
+        TypeResolution resolution = isNumeric(left, sourceText(), ParamOrdinal.FIRST);
+        if (resolution.unresolved()) {
+            return resolution;
+
+        }
+
+        return right == null ? TypeResolution.TYPE_RESOLVED : isNumeric(right, sourceText(), ParamOrdinal.SECOND);
+    }
+
+    @Override
+    protected Pipe makePipe() {
+        return new BinaryOptionalMathPipe(source(), this,
+            Expressions.pipe(left),
+            right == null ? null : Expressions.pipe(right),
+            operation());
+    }
+    
+    protected abstract BinaryOptionalMathOperation operation();
+
+    @Override
+    public boolean foldable() {
+        return left.foldable()
+                && (right == null || right.foldable());
+    }
+
+    @Override
+    public Object fold() {
+        return operation().apply((Number) left.fold(), (right == null ? null : (Number) right.fold()));
+    }
+    
+    @Override
+    public ScriptTemplate asScript() {
+        ScriptTemplate leftScript = asScript(left);
+        ScriptTemplate rightScript = right == null ? null : asScript(right);
+
+        return asScriptFrom(leftScript, rightScript);
+    }
+
+    private ScriptTemplate asScriptFrom(ScriptTemplate leftScript, ScriptTemplate rightScript) {
+        if (right == null) {
```

Comment:
Wouldn't be possible to avoid the "double" methods in painless and just have one with both args where null is passed for the right arg?

Reply:
Gave it a try, but failed. If you have any ideas, I'd love to hear them. Otherwise, I'll try later today or tomorrow.

Reply:
I took some time and it works like that: https://github.com/matriv/elasticsearch/commit/b35090e91f54f7db00e916e4671ad00fcd551cf4

Reply:
Indeed. Nice one. I've incorporated it in the PR. Thanks.